### PR TITLE
Add CMakeLists.txt to build mp11 as standalone library

### PR DESCRIPTION
**Pull request by oliora**
Note that mp11 still requires core Boost headers but can be build against
older version of Boost

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+
+include(GNUInstallDirs)
+include(CMakePackageConfigHelpers)
+
+enable_testing()
+
+find_package(Boost 1.56 REQUIRED)  # 1.56+ for lightweight_test
```
Comment (ax3l) [inline CMakeLists.txt]: is this only required for tests?

if so, you can react on the user-controllable var `BUILD_TESTING[=ON]` which is set by `enable_testing()` and only depend on Boost when testing is requested.

Comment (oliora) [reply CMakeLists.txt]: Mp11 depends on `boost/config.hpp` header so I can't cut dependency on Boost.

**Comment by oliora** [comment]
@pdimov any comment on my PR?

**Comment by pdimov** [comment]
I no longer support standalone use, at least not officially, so I'm not sure about this. Supporting an older Boost requires f.ex. switching back from `config/workaround.hpp` to the deprecated `detail/workaround.hpp`. Same if I decide to make use of a new Boost.Config macro such as f.ex. `BOOST_NO_CXX17_FOLD_EXPRESSIONS` - it won't be present in an older Boost.

**Comment by oliora** [comment]
@pdimov Makes sense but pity. It would be awesome to have mp11 as a standalone library especially with C++17 where most of vocabulary types like `variant` are in the STL so much more things can be done without depending on Boost (one is unfortunately a quite heavy dependency)

**Comment by oliora** [comment]
Btw, is there any chance to get mp11 as part of the STL?

**Comment by pdimov** [comment]
I have a proposal, http://open-std.org/JTC1/SC22/WG21/docs/papers/2018/p0949r0.html, but the committee isn't sure yet what metaprogramming facilities it wants. They'll need to decide one way or the other at the next meeting.

File: CMakeLists.txt (hunk 1 above)
Comment (pdimov) [reply CMakeLists.txt]: I removed the dependency on Config by the way.

Comment (oliora) [reply CMakeLists.txt]: I'll update the PR as @ax3l suggested

Comment (ax3l) [reply CMakeLists.txt]: @oliora do you want to update the PR description accordingly? :)

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,49 @@
+if(APPLE)
+    cmake_minimum_required(VERSION 3.5)
```
Comment (pdimov) [inline CMakeLists.txt]: Out of curiosity, why is this different on Mac?

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.2)
+cmake_minimum_required(VERSION 3.1)
```
Comment (oliora) [inline CMakeLists.txt]: Since we don't use any CMake 3.2 feature, we can relax the required version to 3.1. 3.1. is the version where `CMAKE_CXX_STANDARD` variable was introduced.

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+project(BoostMp11 VERSION 1.66.0 LANGUAGES CXX)
```
Comment (oliora) [inline CMakeLists.txt]: The version should be probably 1.68 already. Time flies :)

File: CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
```
Comment (oliora) [inline CMakeLists.txt]: Since mp11 can optionally use features from C++17 and does its own C++ feature detection it's probably better to set `CMAKE_CXX_STANDARD_REQUIRED` to `OFF` and set `CMAKE_CXX_STANDARD` into the max standard supported by CMake. I'd prefer to make it in a separate commit if possible.

Comment (ax3l) [reply CMakeLists.txt]: If you set `CMAKE_CXX_STANDARD` to the max know to CMake you will exclude a lot of compilers, such as NVCC, that do not yet have e.g. C++17 support.

Comment (oliora) [reply CMakeLists.txt]: Yep. But I just realized that the C++ standard is really applied only to tests since the library itself is header-only. For tests it would be the best to test different language versions to make sure that mp11 compatibility layer works fine and library is compatible with all C++ versions (C++11, C++14 and C++17 for now). That's quite easy to add and I'll do it later in separate PR.

Comment (ax3l) [reply CMakeLists.txt]: Yes, in that case only set the requirement on the tests (per target) instead of putting it on the library target :)

Comment (oliora) [reply CMakeLists.txt]: Let's keep it like it is for this PR. I'd prefer to move c++11 requirement to test targets in a separate commit.

File: CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+add_library(mp11 INTERFACE)
+add_library(Boost::mp11 ALIAS mp11)
+
+target_include_directories(mp11 INTERFACE
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+    $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
+)
+
+if (BUILD_TESTING)
+    find_package(Boost 1.56 REQUIRED)  # 1.56+ for lightweight_test
```
Comment (pdimov) [inline CMakeLists.txt]: Travis of course has 1.55. FetchContent is best, but that's CMake 3.11, and Travis has 3.9. It ain't easy.

Comment (pdimov) [reply CMakeLists.txt]: Ordinarily I'd just submodule Config/Core/Assert, but it's not appropriate for a Boost library to do that.

**Comment by pdimov** [comment]
Thank you for this pull request. I've added a `CMakeLists.txt` that is roughly functionally identical to your proposed one, with certain minor differences that are expected to better fit the overall Boost CMake structure if/when that arrives.

I've left the `export(EXPORT ...)` part commented out for now until I figure out if I need it and for what it might be useful. -)

It's possible to `find_package(Boost 1.56)` first and then fetch as a fallback, but since Travis doesn't have Boost 1.56, there's no immediate need for that. :-)

My next task here would probably be to support building the documentation with `cmake --build . --target doc`, and perhaps install it as well.

If what I added strikes you as wrong, please let me know. I'm not a CMake expert by any means.

**Comment by oliora** [comment]
@pdimov the added CMake files look good for me. I’m closing my PR then.
